toc/rfc/rfc-draft-add-support-for-file-based-service-binding.md

```diff
@@ -0,0 +1,111 @@
+   * Advantages:
+      * CF community could re-use service binding libraries from the Cloud Native community
+      * Moving application between CF or K8s deployments will be easier
+   * Disadvantages
+      * Higher implementation efforts for the Cloud Controller
+
+
+The 2) alternatives offers more than just addressing the issue of this RFC. It suggests an option to evolve the CF platform towards a different service binding specification defined by the Cloud Native community. This means higher implementation efforts for the CF platform and application developers, but possible benefits from the Cloud Native community. This RFC has a light preference for the 2) alternative because of the listed advantages but the feedback of the CF community is wanted here.
+
+Additionally, the application environment is stored in the `CCDB` and `BBS DB` that is why we should define a limit for the size of it, which makes it possible to be stored in the according DBs and doesn’t impact the performance of the communication between Cloud Controller and the Diego API. That is why this RFC suggests a limit of `1MB`, which is roughly ten times higher than the current one of 130KB. This is subject for evaluation during the implation of this RFC.
+
+### Implementation Overview
+
+Cloud Controller should introduce a new app feature for activating the file-based approach. This means that the [App Features API](https://v3-apidocs.cloudfoundry.org/version/3.159.0/index.html#app-features) could be used here and a new feature flag called “file-based-service-bindings" should be introduced.
+The [contract](https://github.com/cloudfoundry/bbs/blob/main/doc/actions.md) between Cloud Controller and Diego should be extended so that file name and file content for the application container can be specified. E.g. the run action could look like this when file approach is selected:
+
+```
+action := &models.RunAction{
+  Path: "/path/to/executable",
+  Args: []string{"some", "args to", "pass in"},
+  Dir: "/path/to/working/directory",
+  User: "username",
+  EnvironmentVariables: []*models.EnvironmentVariable{
+    {
+      Name: "ENVNAME",
+      Value: "ENVVALUE",
+    },
+  },
+ Files: []*models.Files{
+    {
+      Name: "/etc/cf-instance-binding",
+      Value: "VALUE",
+    },
+  },
+  ResourceLimits: &models.ResourceLimits{
+    Nofile: 1000,
+  },
+  LogSource: "some-log-source",
+  SuppressLogOutput: false,
+}
+```
+
+## Workstreams
+
+### App Runtime Interfaces WG
+
+Cloud Controller should be extended with an App Feature to activate the new file base Service Binding option. If the file-based service binding feature is active the Cloud Controller should generate a Run action, which configures the Service Bindings to be stored as tmpfs file(s) in the application container instead of `VCAP_SERVICES` environment variable.
+
+Additionally, the suggest limit for the size should be implemented.
```

Comment:
It would be good to add a section describing how the `name` and `tags` [OSBAPI fields](https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#service-offering-object) could be mapped to `type` and `provider` [service binding fields](https://servicebinding.io/spec/core/1.1.0/#example-directory-structure) (if these are not specified by the broker). This could potentially help with initial broker compatibility. This could for example be done based on well known tags like `postgres` and `mysql`. This would be in-line with the[ service-specific code that is currently used to populate the DATABASE_URL variable](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/lib/cloud_controller/database_uri_generator.rb#L3-L7).

Reply:
In libcnb for Cloud-Native Buildpacks, we implemented a translation from `VCAP_SERVICES` to CNB bindings. What we did was this:

1. Credentials map was turned into key/value files (i.e. key is the file name, value is the file contents). We did not go into nested maps, so if a key in Credentials had a nested JSON value, then the JSON is written as the contents of the file.
2. We used the Label from `VCAP_SERVICES` as the `type`.
3. We used the top-level key from the `VCAP_SERVICES` map as the `provider`.

I haven't used it myself, but I know people are using it. @loewenstein - How has this mapping worked out in practice?

Reply:
We contributed that to `libcnb` mainly because of the Paketo Dynatrace Buildpack and for that it works - as far as I can tell.

Reply:
@loewenstein is https://github.com/buildpacks/libcnb/pull/228 you are referring to? The mentioned mapping above isn't directly visible in the change but most probably I have to dive deeper into it.

Reply:
Looking into the implementation of the CC the top-level key from the `VCAP_SERVICES` [seems to be also the label](https://github.com/cloudfoundry/cloud_controller_ng/blob/b8c6dbef8bc469d0f0dca6de77a482b2038b3dac/app/presenters/system_environment/system_env_presenter.rb#L15-L22) from the `VCAP_SERVICES`. The value is taken from the `name` attribute of the service offering as defined in the OSBAPI [spec](https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#catalog-management). I think it make sense to be consistent and have the same mapping as implemented in `libcnb` and will update the RFC with this if we agree to go with the 2) approach and don't here anything against the suggested mapping.

Reply:
I added details about the `VCAP_SERVICES` -> K8s bindings mapping with https://github.com/cloudfoundry/community/pull/804/commits/88f5fb6e54b3c292396d5b56157821d5b7133e04. Please review and provide feedback.